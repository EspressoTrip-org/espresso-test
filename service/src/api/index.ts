export * as organization from './organizations';
export * as compilation from './compilation';
export * as policies from './policies';
export * as tokens from './tokens';
export * as roles from './roles';
export * as users from './users';
