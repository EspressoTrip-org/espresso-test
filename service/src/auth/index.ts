export * from './generated-policies';
export * from './authorizers';
export * from './context';
export * from './viewer';
