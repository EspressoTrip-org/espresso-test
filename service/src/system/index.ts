import { InterServiceClient } from '@journeyapps-platform/sdk-inter-service';
import * as entity_resolver from '@journeyapps-platform/recon-entity-resolver';
import * as invalidator from '@journeyapps-platform/recon-event-invalidator';
import * as micro from '@journeyapps-platform/micro';
import * as adapters from '../adapters';
import * as kafka_utils from './kafka';
import * as resolver from '../recon';
import { MongoDB } from './mongo';
import * as kafka from 'kafkajs';
import ENV from '../env';

export type CreateSystemParams = {
  mocks: string[];
  accounts: {
    endpoint: string;
    client_id: string;
    client_secret: string;
  };
  mongo: micro.mongo.MongoDBAdapterOptions;
  kafka: micro.kafka.KafkaConnectionOptions;
  node_env: string;
};

export class System extends micro.system.MicroSystem {
  context: CreateSystemParams;

  mongo: MongoDB;
  kafka: micro.kafka.KafkaClientProvisioner;
  producer: kafka.Producer;
  resolver: entity_resolver.EntityResolver;

  policy_cache: adapters.policy_cache.PolicyCache;

  constructor(params: CreateSystemParams) {
    super();

    this.context = params;

    const inter_service = new InterServiceClient(params.accounts);

    this.mongo = this.withLifecycle(new MongoDB(params.mongo), {
      start(component) {
        return component.connect();
      },
      stop(component) {
        return component.close();
      }
    });

    this.kafka = this.withLifecycle(micro.kafka.createClientProvisioner(params.kafka), {
      start(component) {
        return kafka_utils.ensureTopics({
          kafka: component(),
          production: params.node_env === 'production'
        });
      }
    });

    this.producer = this.withLifecycle(
      this.kafka('producer').producer({
        createPartitioner: kafka.Partitioners.LegacyPartitioner
      }),
      {
        start(component) {
          return component.connect();
        },
        stop(component) {
          return component.disconnect();
        }
      }
    );

    this.resolver = this.withLifecycle(
      resolver.createResolver(
        this.mongo,
        params.mocks.includes('recon')
          ? undefined
          : {
              external: {
                endpoint: params.accounts.endpoint,
                client: inter_service
              }
            }
      ),
      {}
    );

    this.withLifecycle(
      invalidator.createEventInvalidator({
        resolver: this.resolver,
        kafka: this.kafka('invalidator'),
        modules: [invalidator.resource_operation_events_module, invalidator.cardinal_module]
      }),
      {
        start: async (invalidator) => {
          await invalidator.start();
          await this.resolver.init();
        },
        stop(invalidator) {
          return invalidator.stop();
        }
      }
    );

    this.policy_cache = adapters.policy_cache.createInMemoryPolicyCache();
  }

  static fromENV(env: typeof ENV) {
    return new System({
      mocks: env.MOCK,
      accounts: {
        endpoint: env.JWT_ISSUER,
        client_id: env.ACCOUNTS_CLIENT_ID,
        client_secret: env.ACCOUNTS_CLIENT_SECRET
      },
      mongo: {
        uri: env.MONGO_URI,
        database: env.MONGO_DATABASE,
        username: env.MONGO_USERNAME,
        password: env.MONGO_PASSWORD
      },
      kafka: {
        brokers: env.KAFKA_BROKERS,
        username: env.KAFKA_USERNAME,
        password: env.KAFKA_PASSWORD,
        mechanism: env.KAFKA_MECHANISM as any
      },
      node_env: env.NODE_ENV
    });
  }
}
