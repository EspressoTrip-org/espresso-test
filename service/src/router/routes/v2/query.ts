import * as cardinal from '@journeyapps-platform/types-cardinal';
import * as micro from '@journeyapps-platform/micro';
import { Router } from '../../router';
import * as errors from '../../../errors';
import * as api from '../../../api';
import { CardinalModel, TokenModelAction, UserModelAction } from '@journeyapps-platform/cardinal-catalog';

export const policies_for_user = Router.post(cardinal.ROUTES_V2.QUERY_POLICIES_FOR_USER, {
  validator: micro.schema.createTsCodecValidator(cardinal.QueryPoliciesForUser),
  authorize: async ({ params, context }) => {
    return context.viewer.can(UserModelAction.READ, CardinalModel.USER, params.user_id);
  },
  handler: ({ params, context }) => {
    return api.policies.getPoliciesForUser(context.system, {
      user_id: params.user_id,
      policies: context.viewer.policies
    });
  }
});

export const policies_for_token = Router.post(cardinal.ROUTES_V2.QUERY_POLICIES_FOR_TOKEN, {
  validator: micro.schema.createTsCodecValidator(cardinal.QueryPoliciesForToken),
  authorize: async ({ params, context }) => {
    const token = await api.tokens.getTokenByValue(context.system.mongo, params.token);
    if (!token) {
      throw new errors.Error404('token', '<redacted>');
    }
    return context.viewer.can(TokenModelAction.READ, CardinalModel.TOKEN, token.id);
  },
  handler: async ({ params, context }) => {
    const token = await api.tokens.getTokenByValue(context.system.mongo, params.token);
    if (!token) {
      throw new errors.Error404('token', '<redacted>');
    }
    return api.policies.getPoliciesForToken(context.system, {
      token_id: token.id,
      policies: context.viewer.policies
    });
  }
});

export const routes = [policies_for_user, policies_for_token];
