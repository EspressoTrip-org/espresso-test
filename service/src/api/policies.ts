import * as cardinal from '@journeyapps-platform/types-cardinal';
import * as micro from '@journeyapps-platform/micro';
import * as compilation from './compilation';
import { stripTokenValue } from './tokens';
import { MongoDB } from '../system/mongo';
import * as producer from './producer';
import * as query_utils from './utils';
import * as crypto from 'node:crypto';
import { System } from '../system';
import * as tokens from './tokens';
import * as mongo from 'mongodb';
import * as users from './users';
import * as roles from './roles';
import * as auth from '../auth';
import * as bson from 'bson';
import { CardinalModel } from '@journeyapps-platform/cardinal-catalog';
import { WithActor } from './utils';

export const getPolicy = async (mongo: MongoDB, id: string) => {
  return query_utils.getByIdOrThrow(mongo.policies, id);
};

type CreatePolicyParams = WithActor<{
  policy: cardinal.Policy & cardinal.Managed;
}>;

export const createPolicy = async (system: System, params: CreatePolicyParams) => {
  const { policy } = params;

  const now = new Date();

  const id = new bson.ObjectId();
  await system.mongo.policies.insertOne({
    _id: id,
    description: policy.description,
    org_id: policy.org_id,
    name: policy.name,
    statements: policy.statements,
    managed: policy.managed,
    created_at: now,
    updated_at: now
  });

  const new_policy = cardinal.PolicyResource.encode({
    _id: id,
    ...policy,
    created_at: now,
    updated_at: now
  });

  await producer.produceAuthEvents(system, {
    actor: params.actor,
    events: {
      type: cardinal.PolicyAuthEventType.POLICY_CREATED,
      payload: new_policy
    }
  });

  return new_policy;
};

export type UpdatePolicyParams = WithActor<{
  policy: cardinal.Managed & cardinal.UpdatePolicyParams;
}>;

export const updatePolicy = async (system: System, params: UpdatePolicyParams) => {
  const { policy } = params;

  const existing_policy = await getPolicy(system.mongo, params.policy.id);

  const now = new Date();
  const update = {
    name: policy.name,
    description: policy.description,
    statements: policy.statements,
    updated_at: now
  };

  await system.mongo.policies.updateOne(
    {
      _id: new bson.ObjectId(policy.id)
    },
    {
      $set: update
    }
  );

  const new_policy = {
    ...existing_policy,
    ...update
  };

  await producer.produceAuthEvents(system, {
    actor: params.actor,
    events: {
      type: cardinal.PolicyAuthEventType.POLICY_UPDATED,
      payload: micro.mongo.toJson(new_policy)
    }
  });

  return new_policy;
};

export const getPolicyAssignments = async (mongo: MongoDB, id: string) => {
  const policy = await getPolicy(mongo, id);

  const filter = {
    policy_ids: {
      $in: [policy.id]
    }
  };

  const users = await mongo.users.find(filter).toArray();
  const tokens = await mongo.tokens.find(filter).toArray();
  const roles = await mongo.roles.find(filter).toArray();

  return {
    users: users,
    tokens: tokens.map(stripTokenValue),
    roles: roles
  };
};

export const deletePolicy = async (system: System, params: WithActor<{ id: string }>) => {
  const policy = await getPolicy(system.mongo, params.id);

  const filter: mongo.Filter<{ policy_ids?: string[] }> = {
    policy_ids: {
      $in: [policy.id]
    }
  };

  // We first find all resources with associations so that we can emit
  // events for them changing
  const { users, tokens, roles } = await getPolicyAssignments(system.mongo, params.id);

  await system.mongo.policies.deleteOne({
    _id: new bson.ObjectId(params.id)
  });

  const update: mongo.UpdateFilter<{ policy_ids: string[] }> = {
    $pull: {
      policy_ids: policy.id
    }
  };

  // Update all associated resources, removing the policy id from their
  // set
  await system.mongo.users.updateMany(filter, update);
  await system.mongo.tokens.updateMany(filter, update);
  await system.mongo.roles.updateMany(filter, update);

  const user_auth_events: cardinal.AuthEvent[] = users.map(micro.mongo.toJson).map((user) => {
    return {
      type: cardinal.UserAuthEventType.USER_ASSIGNMENTS_CHANGED,
      payload: {
        ...user,
        policy_ids: user.policy_ids?.filter((id) => id !== policy.id)
      }
    };
  });

  const token_auth_events: cardinal.AuthEvent[] = tokens.map(micro.mongo.toJson).map((token) => {
    return {
      type: cardinal.TokenAuthEventType.TOKEN_UPDATED,
      payload: {
        ...token,
        policy_ids: token.policy_ids?.filter((id) => id !== policy.id)
      }
    };
  });

  const role_auth_events: cardinal.AuthEvent[] = roles.map(micro.mongo.toJson).map((role) => {
    return {
      type: cardinal.RoleAuthEventType.ROLE_UPDATED,
      payload: {
        ...role,
        policy_ids: role.policy_ids?.filter((id) => id !== policy.id)
      }
    };
  });

  const auth_events = user_auth_events
    .concat(token_auth_events)
    .concat(role_auth_events)
    .concat([
      {
        type: cardinal.PolicyAuthEventType.POLICY_DELETED,
        payload: micro.mongo.toJson(policy)
      }
    ]);

  await producer.produceAuthEvents(system, {
    actor: params.actor,
    events: auth_events
  });
};

export const listUnscopedPolicies = (mongo: MongoDB) => {
  return mongo.policies
    .find({
      org_id: {
        $exists: false
      }
    })
    .map(micro.mongo.toJson)
    .toArray();
};

type ListScopedPoliciesForOrganizationParams = {
  org_id: string;
};

export const listScopedPoliciesForOrganization = async (
  mongo: MongoDB,
  params: ListScopedPoliciesForOrganizationParams
) => {
  return await mongo.policies
    .find({
      org_id: params.org_id
    })
    .map(micro.mongo.toJson)
    .toArray();
};

export const listPolicies = async (
  system: System,
  params: query_utils.RequirePermissions<cardinal.V2ListResourceParams>
) => {
  const query = await query_utils.wrapQueryWithPermissions(system, {
    ...params,

    query: micro.mongo.findBy<cardinal.PolicyResource>({
      id: params.id,
      org_id: params.org_id
    }),
    model: CardinalModel.POLICY
  });

  return await micro.mongo.paginate(system.mongo.policies, {
    query: query,
    cursor: params.cursor,
    limit: params.limit || 100,
    transform: micro.mongo.toJson
  });
};

export const getPoliciesByIds = async (mongo: MongoDB, ids: string[]): Promise<cardinal.SerializedPolicyResource[]> => {
  return mongo.policies.find(micro.mongo.getByIds(ids)).map(micro.mongo.toJson).toArray();
};

export const getPolicyIdsForUser = async (
  system: System,
  params: { user: cardinal.SerializedUserResource }
): Promise<string[]> => {
  const { user } = params;
  const user_roles = user?.role_ids ? await roles.getRoles(system.mongo, user.role_ids) : [];
  return user_roles.reduce((ids: string[], role) => {
    ids.push(...(role.policy_ids || []));
    return ids;
  }, user.policy_ids || []);
};

export const getPoliciesForUser = (system: System, params: query_utils.OptionalPermissions<{ user_id: string }>) => {
  return micro.tracing.trace('get-policies-for-user', async (span) => {
    const mongo = system.mongo;

    const user = await users.getUser(mongo, params.user_id);
    const organization = await mongo.organizations.findOne(micro.mongo.getById(user.org_id));

    if (!organization) {
      throw new micro.errors.ResourceNotFound('organization', user.org_id);
    }

    if (organization.locked) {
      const policy = auth.createLockedPolicy(organization._id.toHexString());
      return [policy];
    }

    let requested_policy_ids = await getPolicyIdsForUser(system, { user });
    let policy_ids = requested_policy_ids;
    if (params.policies) {
      policy_ids = await system.resolver.resolveResourceIds({
        policies: params.policies,
        model: CardinalModel.POLICY,
        entity_filter: {
          id: requested_policy_ids
        }
      });
    }

    const hash = crypto.createHash('sha256');
    hash.update(user.id);
    hash.update(policy_ids.join(''));
    const key = hash.digest('hex');

    const cache_hit = await system.policy_cache.get(key);
    span.setAttribute('cached', !!cache_hit);
    if (cache_hit) {
      return cache_hit;
    }

    const locked_orgs = await micro.tracing.trace('fetch-locked-orgs', async () => {
      const org_ids = await mongo.policies
        .find({
          $and: [
            micro.mongo.getByIds(policy_ids),
            {
              org_id: {
                $exists: true
              }
            }
          ]
        })
        .project<{ org_id: string }>({
          org_id: 1
        })
        .map((policy) => new bson.ObjectId(policy.org_id))
        .toArray();

      return mongo.organizations
        .find({
          _id: {
            $in: org_ids
          },
          locked: true
        })
        .project<{ _id: bson.ObjectId }>({
          _id: 1
        })
        .map((org) => org._id.toHexString())
        .toArray();
    });

    return await micro.tracing.trace('collect-policies', async () => {
      const policies = await mongo.policies
        .find({
          $and: [
            micro.mongo.getByIds(policy_ids),
            {
              org_id: {
                $nin: locked_orgs
              }
            }
          ]
        })
        .project<{ _id: bson.ObjectId; org_id: string; statements: cardinal.PolicyStatement[] }>({
          _id: 1,
          org_id: 1,
          statements: 1
        })
        .map(({ _id, ...policy }) => {
          const _policy = {
            id: _id.toHexString(),
            ...policy
          };
          return compilation.interpolateTemplatePolicy(_policy, {
            actor: {
              id: user.id,
              type: cardinal.ActorType.User,
              org_id: user.org_id
            }
          });
        })
        .toArray();

      await system.policy_cache.set(key, policies);

      return policies;
    });
  });
};

export const getPoliciesForToken = (system: System, params: query_utils.OptionalPermissions<{ token_id: string }>) => {
  return micro.tracing.trace('get-policies-for-token', async () => {
    const token = await tokens.getToken(system.mongo, params.token_id);

    /**
     * If the token is a scoped token, check to see if the organization it belongs to is
     * locked.
     *
     * If it is locked then we don't return policies for the token
     */
    if (token.org_id) {
      const organization = await system.mongo.organizations.findOne(micro.mongo.getById(token.org_id));

      if (!organization) {
        throw new micro.errors.ResourceNotFound('organization', token.org_id);
      }

      if (organization.locked) {
        return [];
      }
    }

    let policy_ids = token.policy_ids || [];
    if (params.policies && token.policy_ids) {
      policy_ids = await system.resolver.resolveResourceIds({
        policies: params.policies,
        model: CardinalModel.POLICY,
        entity_filter: {
          id: token.policy_ids
        }
      });
    }

    if (policy_ids.length > 0) {
      return await micro.tracing.trace('collect-policies', () => {
        return system.mongo.policies
          .find(micro.mongo.getByIds(policy_ids))
          .project<{ _id: bson.ObjectId; statements: cardinal.PolicyStatement[] }>({
            _id: 1,
            org_id: 1,
            statements: 1
          })
          .map(({ _id, ...policy }) => {
            const _policy = {
              id: _id.toHexString(),
              ...policy
            };
            return compilation.interpolateTemplatePolicy(_policy, {
              actor: {
                id: token.id,
                type: cardinal.ActorType.Token,
                org_id: token.org_id
              }
            });
          })
          .toArray();
      });
    }

    return [];
  });
};
