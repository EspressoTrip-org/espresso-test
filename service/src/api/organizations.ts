import * as cardinal from '@journeyapps-platform/types-cardinal';
import * as micro from '@journeyapps-platform/micro';
import * as producer from './producer';
import * as policies from './policies';
import * as tokens from './tokens';
import { System } from '../system';
import * as roles from './roles';
import * as users from './users';
import * as auth from '../auth';
import * as bson from 'bson';

type PolicyProvisionParams = {
  org_id: string;
};

export const provisionStandardPoliciesAndRolesForOrg = async (system: System, params: PolicyProvisionParams) => {
  const actor: cardinal.SystemActor = {
    type: cardinal.ActorType.System
  };

  const generated_policies = auth.createDefaultPoliciesForOrg(params.org_id);
  const new_policies = await Promise.all(
    Object.values(generated_policies).map((policy) => {
      return policies.createPolicy(system, {
        actor: actor,
        policy: {
          ...policy,
          managed: true
        }
      });
    })
  );

  const policy_map = new_policies.reduce((acc: auth.CreateDefaultRolesPolicyMap, policy) => {
    acc[policy.name as auth.MANAGED_POLICY] = policy.id;
    return acc;
  }, {});

  const generated_roles = auth.createDefaultRolesForOrg(params.org_id, policy_map);
  return await Promise.all(
    generated_roles.map((role) => {
      return roles.createRole(system, {
        actor,
        role: {
          ...role,
          managed: true
        }
      });
    })
  );
};

type EmitInvalidationEventParams = {
  org: micro.db.SerializedId & cardinal.Organization;
};

/**
 * Emit user assignment changed events which act as cache invalidation events.
 */
export const emitInvalidationEvents = async (system: System, params: EmitInvalidationEventParams) => {
  const org_users = await users.listUsersForOrganization(system.mongo, {
    org_id: params.org.id
  });

  await producer.produceAuthEvents(system, {
    actor: {
      type: cardinal.ActorType.System
    },
    events: org_users.map((user) => {
      return {
        type: cardinal.UserAuthEventType.USER_ASSIGNMENTS_CHANGED,
        payload: user
      };
    })
  });
};

type UpsertOrganizationParams = {
  org: micro.db.SerializedId & cardinal.Organization;
  created_by?: string;
};

export const upsertOrganization = async (system: System, params: UpsertOrganizationParams) => {
  const existing = await system.mongo.organizations.findOne(micro.mongo.getById(params.org.id));

  await system.mongo.organizations.updateOne(
    {
      _id: new bson.ObjectId(params.org.id)
    },
    {
      $set: {
        name: params.org.name,
        locked: params.org.locked
      }
    },
    { upsert: true }
  );

  if (existing) {
    const locked_changed = params.org.locked !== existing.locked;
    if (locked_changed) {
      await emitInvalidationEvents(system, params);
    }

    return;
  }

  const roles = await provisionStandardPoliciesAndRolesForOrg(system, {
    org_id: params.org.id
  });

  // If there is a created_by field on the event then assign the Owner role to the user who
  // created this organization
  if (!params.created_by) {
    return;
  }

  try {
    const user = await users.getUser(system.mongo, params.created_by);
    const owner_role = roles.find((role) => role.name === cardinal.MANAGED_ROLE.OWNER);
    if (!owner_role) {
      return;
    }

    await users.updateUserAssignment(system, {
      actor: {
        type: cardinal.ActorType.System
      },
      assignment: 'role_ids',
      assignment_id: owner_role.id,
      id: user.id,
      op: 'add'
    });
  } catch (e) {
    micro.logger.error(`user ${params.created_by} (creator of org) could not be found`);
  }
};

export const deleteOrganization = async (system: System, params: { id: string }) => {
  await system.mongo.organizations.deleteOne({
    _id: new bson.ObjectId(params.id)
  });

  const all_policies = await policies.listScopedPoliciesForOrganization(system.mongo, {
    org_id: params.id
  });
  const all_roles = await roles.listRolesByOrganization(system.mongo, {
    org_id: params.id
  });
  const all_tokens = await tokens.listScopedTokensForOrganization(system.mongo, {
    org_id: params.id
  });

  // This is a pretty inefficient way to clean up but org deletion will be very in-frequent
  // and as such this is a nice way to re-use already tested logic to do the cleanup

  for (const token of all_tokens) {
    await tokens.deleteToken(system, {
      actor: {
        type: cardinal.ActorType.System
      },
      id: token.id
    });
  }

  for (const role of all_roles) {
    await roles.deleteRole(system, {
      actor: {
        type: cardinal.ActorType.System
      },
      id: role.id
    });
  }

  for (const policy of all_policies) {
    await policies.deletePolicy(system, {
      actor: {
        type: cardinal.ActorType.System
      },
      id: policy.id
    });
  }
};
