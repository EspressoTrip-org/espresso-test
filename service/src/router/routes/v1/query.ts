// transitive dependencies
import type * as micro from '@journeyapps-platform/micro';

import { Router } from '../../router';
import * as v2 from '../v2/query';

export const policies_for_user = Router.get('/api/v1/users/:user_id/get-policies', {
  validator: v2.policies_for_user.validator,
  authorize: v2.policies_for_user.authorize,
  handler: v2.policies_for_user.handler
});

export const policies_for_token = Router.post('/api/v1/tokens/get-policies', {
  validator: v2.policies_for_token.validator,
  authorize: v2.policies_for_token.authorize,
  handler: v2.policies_for_token.handler
});

export const query_routes = [policies_for_user, policies_for_token];
