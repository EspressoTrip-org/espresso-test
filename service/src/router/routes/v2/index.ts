// transitive dependency
import type * as micro from '@journeyapps-platform/micro';

import * as assignments from './assignments';
import * as policies from './policies';
import * as tokens from './tokens';
import * as roles from './roles';
import * as query from './query';

export const v2_routes = [
  ...assignments.routes,
  ...policies.routes,
  ...roles.routes,
  ...tokens.routes,
  ...query.routes
];
