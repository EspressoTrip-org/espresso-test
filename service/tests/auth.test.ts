import { beforeAll, beforeEach, afterAll, describe, test, expect } from 'vitest';

import * as cardinal from '@journeyapps-platform/types-cardinal';
import * as routes from '../src/router/routes/v2/query';
import { System } from '../src/system/index';
import * as auth from '../src/auth/index';
import * as api from '../src/api/index';
import * as mocks from './mocks/index';
import * as bson from 'bson';

let system: System;

const org_id = '5f2ab36c60d2294576efa646';

beforeAll(async () => {
  system = await mocks.createMockedSystem();
  await system.start();

  await api.organization.upsertOrganization(system, {
    org: {
      id: org_id,
      name: 'test org'
    }
  });
});

afterAll(async () => {
  await system.stop();
});

beforeEach(async () => {
  // @ts-ignore
  await system.mongo.policies.deleteMany({});
  await system.mongo.users.deleteMany({});
});

describe('auth', () => {
  const policy_id = new bson.ObjectId('5e6f5519ede3631176831f4c');
  const now = new Date('2020-03-20T12:31:48.370Z');

  const policy: cardinal.PolicyResource = {
    _id: policy_id,
    name: 'Policy',
    description: 'Some Policy',
    org_id: org_id,
    statements: [],
    created_at: now,
    updated_at: now
  };

  test('can read policies assigned to user', async () => {
    await system.mongo.policies.insertOne(policy);

    const user_id = new bson.ObjectId();
    const user = await system.mongo.users.insertOne({
      _id: user_id,

      email: 'test-user@journeyapps.com',
      policy_ids: [policy_id.toHexString()],
      role_ids: [],
      org_id: org_id
    });

    await system.resolver.init();

    const res = await routes.policies_for_user.handler(
      mocks.createMockedPayload({
        system,
        policies: auth.canReadOwnPolicies({
          policies: await api.policies.getPoliciesForUser(system, { user_id: user_id.toHexString() }),
          model: 'user',
          id: user_id.toHexString(),
          scope: org_id
        }),
        params: {
          user_id: user.insertedId.toHexString()
        }
      })
    );
    expect(res).toMatchSnapshot();
  });

  test('can read policies assigned to token', async () => {
    await system.mongo.policies.insertOne(policy);

    const token = await api.tokens.createToken(system, {
      token: {
        org_id,
        policy_ids: [policy_id.toHexString()]
      },
      actor: {
        type: cardinal.ActorType.System
      }
    });

    await system.resolver.init();

    const res = await routes.policies_for_token.handler(
      mocks.createMockedPayload({
        system,
        policies: auth.canReadOwnPolicies({
          policies: await api.policies.getPoliciesForToken(system, { token_id: token.id }),
          model: 'token',
          id: token.id,
          scope: org_id
        }),
        params: {
          token: token.value
        }
      })
    );
    expect(res).toMatchSnapshot();
  });
});
