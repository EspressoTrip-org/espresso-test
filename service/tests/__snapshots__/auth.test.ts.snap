// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`auth > can read policies assigned to token 1`] = `
[
  {
    "id": "5e6f5519ede3631176831f4c",
    "org_id": "5f2ab36c60d2294576efa646",
    "statements": [],
  },
]
`;

exports[`auth > can read policies assigned to user 1`] = `
[
  {
    "id": "5e6f5519ede3631176831f4c",
    "org_id": "5f2ab36c60d2294576efa646",
    "statements": [],
  },
]
`;
